### ISSUE TITLE
Documented horizontal form-group fieldset markup is invalid HTML

### ISSUE THREAD

**Issue by tmorehouse** [twbs/bootstrap#24938]
The `<legend>` element must be the first child of the `<fieldset>` element, which is not the case in the horizontal form-group layout using a fieldset:
- Docs: https://getbootstrap.com/docs/4.0/components/forms/#horizontal-form
- HTML5 validation: https://validator.w3.org/nu/?doc=https%3A%2F%2Fgetbootstrap.com%2Fdocs%2F4.0%2Fcomponents%2Fforms%2F

>![image](https://user-images.githubusercontent.com/2781561/33532983-8e805bc8-d874-11e7-8978-6744ca2680cf.png)

```html
<fieldset class="form-group">
  <div class="row">
    <legend class="col-form-legend col-sm-2">Radios</legend>
    <div class="col-sm-10">
      <div class="form-check">
        <label class="form-check-label">
          <input class="form-check-input" type="radio" name="gridRadios" id="gridRadios1" value="option1" checked>
          Option one is this and that&mdash;be sure to include why it's great
        </label>
      </div>
      <div class="form-check">
        <label class="form-check-label">
          <input class="form-check-input" type="radio" name="gridRadios" id="gridRadios2" value="option2">
          Option two can be something else and selecting it will deselect option one
        </label>
      </div>
    </div>
  </div>
</fieldset>
```

![image](https://user-images.githubusercontent.com/2781561/33533124-6cf50b9c-d875-11e7-8292-eab696b261a8.png)

Visually, the above layout appears correct, but semantically it is incorrect (and not always announced correctly by screen readers, as they do not detect the `<legend>`).

Changing the markup so that `<legend>` is a direct first child of `<fieldset>` (and moving class `.row` to the fieldset) produces correctly validated HTML5, but the layout breaks:

```html
<fieldset class="form-group row">
  <legend class="col-form-legend col-sm-2">Radios</legend>
  <div class="col-sm-10">
    <div class="form-check">
      <label class="form-check-label">
          <input class="form-check-input" type="radio" name="gridRadios" id="gridRadios1" value="option1" checked>
          Option one is this and that&mdash;be sure to include why it's great
        </label>
    </div>
    <div class="form-check">
      <label class="form-check-label">
          <input class="form-check-input" type="radio" name="gridRadios" id="gridRadios2" value="option2">
          Option two can be something else and selecting it will deselect option one
        </label>
    </div>
  </div>
</fieldset>
```
Produces the following (the legend appears on a "line" by itself, even though its width is correct):

![image](https://user-images.githubusercontent.com/2781561/33533141-9312ddfe-d875-11e7-98c1-ce3dc6bd3e43.png)

After playing with the classes/styles, the only way to get correct semantic markup **_and_** correct visual layout was to float the legend to the left:

```html
<fieldset class="form-group row">
  <legend class="float-left col-form-legend col-sm-2">Radios</legend>
  <div class="col-sm-10">
    <div class="form-check">
      <label class="form-check-label">
          <input class="form-check-input" type="radio" name="gridRadios" id="gridRadios1" value="option1" checked>
          Option one is this and that&mdash;be sure to include why it's great
        </label>
    </div>
    <div class="form-check">
      <label class="form-check-label">
          <input class="form-check-input" type="radio" name="gridRadios" id="gridRadios2" value="option2">
          Option two can be something else and selecting it will deselect option one
        </label>
    </div>
  </div>
</fieldset>
```

![image](https://user-images.githubusercontent.com/2781561/33533173-c5d41208-d875-11e7-8736-802cdcbcb166.png)

Browser used for testing: Firefox
Codepen: https://codepen.io/tmorehouse/pen/BmMXKL

Maybe `float: left;` should be added to the class `.col-form-legend` (now known just as `.col-form-label`) to allow valid HTML5 markup and visually correct layout?

Since the issue is only with the `<label>` element, the following additional CSS could be used as an alternative:

```css
fieldset > legend.col-form-label {
  float: left;
}
```

Testing in mobile chrome, the "float" fix doesn't appear to work.

I am not sure if there is any alternative CSS trickery that could be used to correct this.

p.s. the div surrounding the radio inputs should have ` role="radiogroup"` (or `role="group"` if checkboxes are used)

**Comment by tmorehouse** [comment]
It appears that each browser flavor implements its own rules on how the label of a legend is treated (and in most browsers it is an out-of-flow special element).

The only alternative in dealing with invalid mrakup is to use an `aria-labelledby` attribute on the fieldset pointing to the ID of the `legend`.  While still invalid HTML5 markup, it does provider screen readers context of the fieldset.

**Comment by patrickhlauke** [comment]
> The <legend> element must be the first child of the <fieldset> element, which is not the case in the horizontal form-group layout using a fieldset:

yup, aware of this, and consciously decided to leave as is, since browsers/AT support it in its current form, and layout gets..."tricky"...if trying to do it properly.

and forcing authors to use `aria-labelledby` etc is a bit much just to please the validator. heck, may see if there's any chance in loosening the spec on that aspect

**Comment by tmorehouse** [comment]
The `aria-labelledby` on the `<fieldset>` (pointing to the `<legend>`) doesn't make it pass validation, but it does ensure that screen readers will pick it up better.

**Comment by patrickhlauke** [comment]
> it does ensure that screen readers will pick it up better.

are there current problems in particular browser/AT combinations?

**Comment by tmorehouse** [comment]
I haven't experienced any myself, but I have heard of issues from users saying that they don't hear the "title" of the fieldset.  And some readers (i.e. Windows Eyes) don't read them at all (even when they are placed correctly, unless optionally enabled in W.E. settings).

**Comment by mdo** [comment]
Closing as stale and a won't fix, if I'm reading Patrick's three year old comment right. 😬 This is also already addressed in v5, though there's a bug for Safari: https://getbootstrap.com/docs/5.0/forms/layout/#horizontal-form. Opened #32746.
